torch/distributions/distribution.py

```diff
@@ -161,21 +168,26 @@ def _extended_shape(self, sample_shape=torch.Size()):
             shape = torch.Size((1,))
         return shape
 
-    def _validate_log_prob_arg(self, value):
+    def _validate_sample(self, value):
         """
-        Argument validation for `log_prob` methods. The rightmost dimensions
-        of a value to be scored via `log_prob` must agree with the distribution's
-        batch and event shapes.
+        Argument validation for distribution methods such as `log_prob`,
+        `cdf` and `icdf`. The rightmost dimensions of a value to be
+        scored via these methods must agree with the distribution's batch
+        and event shapes.
 
         Args:
-            value (Tensor or Variable): the tensor whose log probability is to be
-                computed by the `log_prob` method.
+            value (Tensor or Variable): the tensor used as the arguments for
+                these methods
         Raises
             ValueError: when the rightmost dimensions of `value` do not match the
                 distribution's batch and event shapes.
         """
         if not (torch.is_tensor(value) or isinstance(value, Variable)):
-            raise ValueError('The value argument to log_prob must be a Tensor or Variable instance.')
+            raise ValueError('The value argument must be a Tensor or Variable instance.')
+
+        if self._validate_args:
```

Comment:
Aha, this is a very good reason to save `self._validate_args`! cc @lazypanda1

Why not move this check into all locations that call `self._validate_sample()`? That way we could avoid additional overhead of a function call and an `isinstance` call.

Reply:
Are you expecting something like:
```
def log_prob(self, value):
    if self._validate_args:
        if not self.support.check(value).all():
            ...
    # remaining log_prob method
```
?

Reply:
I think it would be cleanest to implement
```py
def _validate_sample(self, value):
    # unconditionally validate sample

def log_prob(self, value):
    if self._validate_args:
        self._validate_sample(value)
    # remaining log_prob method
```
but it is almost as clean to
```py
def _validate_sample(self, value):
    if not self._validate_args:
        return
    # validation logic

def log_prob(self, value):
    self._validate_sample(value)
```

Reply:
Won't this affect the shape validation in the process? Are you suggesting that `self._validate_args` will validate everything (support, type and shape) or nothing at all?

Reply:
Yes I think it's sensible for `self._validate_args` to determine both shape validation and value validation. We'll probably need to update some of the shape death tests to turn on validation after this change.

I'm particularly concerned with python overhead of torch.distributions. Whereas tensorflow distributions can check types once and then run the compiled graph repeatedly, torch.distributions must perform any validation at every iteration.